tmc/pkg/server/controllers.go

```diff
@@ -0,0 +1,406 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"fmt"
+	_ "net/http/pprof"
+
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	kcpkubernetesclientset "github.com/kcp-dev/client-go/kubernetes"
+
+	kcpapiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/kcp/clientset/versioned"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/cluster"
+	"github.com/kcp-dev/kcp/pkg/reconciler/apis/apiresource"
+	schedulinglocationstatus "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/location"
+	schedulingplacement "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/placement"
+	workloadsapiexport "github.com/kcp-dev/kcp/pkg/reconciler/workload/apiexport"
+	workloadsapiexportcreate "github.com/kcp-dev/kcp/pkg/reconciler/workload/apiexportcreate"
+	"github.com/kcp-dev/kcp/pkg/reconciler/workload/heartbeat"
+	workloadnamespace "github.com/kcp-dev/kcp/pkg/reconciler/workload/namespace"
+	workloadplacement "github.com/kcp-dev/kcp/pkg/reconciler/workload/placement"
+	workloadresource "github.com/kcp-dev/kcp/pkg/reconciler/workload/resource"
+	synctargetcontroller "github.com/kcp-dev/kcp/pkg/reconciler/workload/synctarget"
+	"github.com/kcp-dev/kcp/pkg/reconciler/workload/synctargetexports"
+)
+
+func postStartHookName(controllerName string) string {
+	return fmt.Sprintf("kcp-tmc-start-%s", controllerName)
+}
+
+func (s *Server) installWorkloadResourceScheduler(ctx context.Context, config *rest.Config) error {
+	config = rest.CopyConfig(config)
+	config = rest.AddUserAgent(config, workloadresource.ControllerName)
+	dynamicClusterClient, err := kcpdynamic.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	resourceScheduler, err := workloadresource.NewController(
+		dynamicClusterClient,
+		s.Core.DiscoveringDynamicSharedInformerFactory,
+		s.Core.KcpSharedInformerFactory.Workload().V1alpha1().SyncTargets(),
+		s.Core.KubeSharedInformerFactory.Core().V1().Namespaces(),
```

Comment:
Nit: Move this one above to put all kcp ones together?

Reply:
It's pure move right now. Cleanups can come in follow-up.